src/web/web_sockets.cpp

```diff
@@ -0,0 +1,560 @@
+        encode_hash,
+        decode_block_header
+    };
+}
+
+web_sockets::~web_sockets()
+{
+    unregister_connections();
+    mg_mgr_free(&manager_);
+}
+
+bool web_sockets::start_web(const config::endpoint& endpoint,
+    bool secure, std::string name)
+{
+    struct mg_bind_opts bind_opts{};
+    const auto security = secure ? "secure" : "public";
+
+    const auto certificate = settings_.web_sockets_ssl_certificate;
+    const auto key = settings_.web_sockets_ssl_key;
+    if (secure && (certificate.empty() || key.empty()))
+    {
+        LOG_ERROR(LOG_SERVER)
+            << "Required SSL certificate and key is not specified.";
+
+        return false;
+    }
+
+    if (settings_.web_sockets_root_path.empty())
+    {
+        LOG_ERROR(LOG_SERVER)
+            << "Required web_sockets_root_path is not specified.";
+
+        return false;
+    }
+
+    mg_mgr_init(&manager_, nullptr);
+    std::memset(&bind_opts, 0, sizeof(bind_opts));
+
+    const auto port = std::to_string(endpoint.port());
+
+    if (secure)
+    {
+        bind_opts.ssl_cert = certificate.c_str();
+        bind_opts.ssl_key = key.c_str();
+
+        LOG_INFO(LOG_SERVER)
+            << "Bound " << security << " " << name
+            << " websocket socket to port " << port
+            << " using certificate " << certificate
+            << " and key " << key;
+    }
+    else
+    {
+        LOG_INFO(LOG_SERVER)
+            << "Bound " << security << " " << name
+            << " websocket socket to port " << port;
+    }
+
+    mg_connection_ptr connection = mg_bind_opt(
+        &manager_, port.c_str(), web_event_handler, bind_opts);
+
+    if (connection == nullptr)
+    {
+        LOG_ERROR(LOG_SERVER)
+            << "Failed to bind websocket listener socket to port "
+            << port << ".";
+
+        return false;
+    }
+
+    connection->user_data = static_cast<void*>(this);
+
+    mg_set_protocol_http_websocket(connection);
+
+    web_socket_options.document_root =
+        settings_.web_sockets_root_path.c_str();
+    web_socket_options.enable_directory_listing = "no";
+
+    return true;
+}
+
+void web_sockets::poll(size_t timeout_milliseconds)
+{
+    const auto deadline = steady_clock::now() +
+        milliseconds(timeout_milliseconds);
+
+    while (steady_clock::now() < deadline)
+        mg_mgr_poll(&manager_, poll_interval);
+}
+
+bool web_sockets::have_websocket_connections()
+{
+    return !connections_.empty();
+}
+
+size_t web_sockets::connection_count()
+{
+    return connections_.size();
+}
+
+bool web_sockets::is_websocket_id(uint32_t id)
+{
+    return id & websocket_id_mask;
+}
+
+bool web_sockets::have_pending_query_work()
+{
+    return !pending_query_work_list_.empty();
+}
+
+void web_sockets::register_connection(mg_connection_ptr connection)
+{
+    connections_.insert(connection);
+}
+
+void web_sockets::unregister_connection(mg_connection_ptr connection)
+{
+    auto connection_iter = connections_.find(connection);
```

Comment:
Just use `it` here (not `iter`) and no need to predicate it with "connection" since it is unambiguous.